### PR TITLE
Document check copies

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR document a little bit better how or `Copied from` framework works, adds comments in the actual scripts and rework a bit the test to be better.

In passing I added a feature requested which was to make sure `make fix-copies` took the function definition or the superclass into account: currently it ignore the whole first line, but if we change the signature of a function / the superclass of a class which is copied from, that modification is not propagated (cc @Rocketknight1 who last requested it)

As you can see from the diff, that feature was direly needed... I had to add `BartPreTrainedModel` (right spelling to be consistent with other models) or break multiple copies, and you can see a lot of signatures or copied from statement being fixed.

### CODE REVIEW

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -502,7 +502,7 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         return hidden_states
 
 
-class BartPretrainedModel(PreTrainedModel):
+class BartPreTrainedModel(PreTrainedModel):
```
Comment (sgugger) [inline src/transformers/models/bart/modeling_bart.py:505]: Switching to the proper name with capital T, and creating an alias for backward compatibility.

File: tests/repo_utils/test_check_copies.py [hunk 2]
```diff
@@ -49,78 +49,138 @@ def forward(self, hidden_states):
...
+    with open(filename, "r", encoding="utf-8") as f:
+        content = f.read()
+
+    content = content.replace(old, new)
+
+    with open(filename, "w", encoding="utf-8") as f:
+        f.write(content)
+
+
+def create_tmp_repo(tmp_dir):
```
Comment (sgugger) [inline tests/repo_utils/test_check_copies.py:108]: Using the same logic as the test fetcher here to create a fake repo and have more interesting tests.

**Review by ydshieh — APPROVED** [review]
Great fix! Thanks.

Just wondering if we have `before running this command, stash or commit your changes first` somewhere in the doc or tutorial (of contribution).

File: src/transformers/models/albert/modeling_flax_albert.py [hunk 3]
```diff
@@ -173,7 +173,6 @@ def setup(self):
         self.LayerNorm = nn.LayerNorm(epsilon=self.config.layer_norm_eps, dtype=self.dtype)
         self.dropout = nn.Dropout(rate=self.config.hidden_dropout_prob)
 
-    # Copied from transformers.models.bert.modeling_flax_bert.FlaxBertEmbeddings.__call__
```
Comment (sgugger) [inline src/transformers/models/albert/modeling_flax_albert.py:176]: @sanchit-gandhi I have to break the connection here as the same method in BERT as `attention_mask` as an argument but adding it here breaks a lot of tests.

File: src/transformers/models/pegasus/modeling_flax_pegasus.py [hunk 4]
```diff
@@ -223,7 +223,7 @@ def shift_tokens_right(input_ids: np.array, pad_token_id: int, decoder_start_tok
 
 
 # Copied from transformers.models.marian.modeling_flax_marian.create_sinusoidal_positions
-def create_sinusoidal_positions(n_pos, dim, dtype):
```
Comment (sgugger) [inline src/transformers/models/pegasus/modeling_flax_pegasus.py:226]: This function is copied from and does not do anything with `dtype` @sanchit-gandhi so adapted its two use below in the file.

File: src/transformers/models/albert/modeling_flax_albert.py (hunk 3 above)
Comment (sanchit-gandhi) [reply src/transformers/models/albert/modeling_flax_albert.py:176]: Good for me - the embeddings don't use an attention mask for Albert (it's instead masked in the attention), so fine to break the connection here!

File: docs/source/en/pr_checks.md [hunk 5]
```diff
@@ -142,3 +142,58 @@ Additional checks concern PRs that add new models, mainly that:
 - All checkpoints used actually exist on the Hub
 
 -->
+
+### Check copies
```
Comment (LysandreJik) [inline docs/source/en/pr_checks.md:146]: Nice doc, thanks!

**Review by LysandreJik — APPROVED** [review]
Very nice, it was indeed needed given the changes. Thanks for adding the much needed documentation, should make it simpler for others to understand the content. 

LGTM

**Comment by Rocketknight1** [comment]
Little late but great fix and the diff is really nice too!
